src/Cassandra/Serialization/Graph/GraphSON2/GraphTypeSerializer.cs

```diff
@@ -0,0 +1,418 @@
+            try
+            {
+                return (T)result;
+            }
+            catch (Exception ex)
+            {
+                var message = result == null
+                    ? $"It is not possible to convert NULL to target type {type.FullName}"
+                    : $"It is not possible to convert type {result.GetType().FullName} to target type {type.FullName}";
+
+                throw new InvalidOperationException(message, ex);
+            }
+        }
+
+        /// <inheritdoc />
+        public object FromDb(JToken token, Type type)
+        {
+            if (TryDeserialize(token, type, DeserializeGraphNodes, out var result))
+            {
+                return result;
+            }
+
+            // No converter is available but the types don't match, so attempt to do:
+            //     (TFieldOrProp) row.GetValue<T>(columnIndex);
+            try
+            {
+                var expr = (ConstantExpression)Expression.Constant(result);
+                var convert = Expression.Convert(expr, type);
+                return Expression.Lambda(convert).Compile().DynamicInvoke();
+            }
+            catch (Exception ex)
+            {
+                var message = result == null
+                    ? $"It is not possible to convert NULL to target type {type.FullName}"
+                    : $"It is not possible to convert type {result.GetType().FullName} to target type {type.FullName}";
+                throw new InvalidOperationException(message, ex);
+            }
+        }
+
+        private bool TryDeserialize(JToken token, Type type, bool useGraphNodes, out dynamic result)
+        {
+            if ((type == typeof(object) && useGraphNodes) || type == typeof(GraphNode) || type == typeof(IGraphNode))
+            {
+                result = new GraphNode(new GraphSONNode(this, token));
+                return true;
+            }
+
+            if (token is JValue)
+            {
+                return ConvertFromDb(_reader.ToObject(token), type, out result);
+            }
+
+            var typeName = string.Empty;
+            if (token is JObject)
+            {
+                typeName = ((string) token[GraphSONTokens.TypeKey]) ?? string.Empty;
+            }
+
+            if (TryConvertFromListOrSet(token, type, typeName, out result))
+            {
+                return true;
+            }
+
+            if (TryConvertFromMap(token, type, typeName, out result))
+            {
+                return true;
+            }
+
+            if (TryConvertFromBulkSet(token, type, typeName, out result))
+            {
+                return true;
+            }
+
+            return ConvertFromDb(_reader.ToObject(token), type, out result);
+        }
+
+        private bool TryConvertFromListOrSet(JToken token, Type type, string typeName, out dynamic result)
+        {
+            if (token is JArray || typeName.Equals("g:List") || typeName.Equals("g:Set"))
```

Comment:
Maybe we can invert ifs on `TryConvertFrom` methods to reduce nesting.

Reply:
that makes perfect sense, I don't know why I did it this way in the first place 😅 

done